src/codecs/avif/yuv.rs

```diff
@@ -0,0 +1,1403 @@
+                } else {
+                    unreachable!();
+                }
+            }
+
+            // Process remainder if width is odd.
+            if image.width & 1 != 0 {
+                let y_left = y_src.chunks_exact(2).remainder();
+                let rgb_chunks = rgb
+                    .chunks_exact_mut(CHANNELS * 2)
+                    .into_remainder()
+                    .chunks_exact_mut(CHANNELS);
+                let u_iter = u_src.iter().rev();
+                let v_iter = v_src.iter().rev();
+
+                for (((y_src, u_src), v_src), rgb_dst) in
+                    y_left.iter().zip(u_iter).zip(v_iter).zip(rgb_chunks)
+                {
+                    let y_value = (y_src.as_() - bias_y) * y_coef;
+                    let cb_value = u_src.as_() - bias_uv;
+                    let cr_value = v_src.as_() - bias_uv;
+
+                    let r = ((y_value + cr_coef * cr_value + ROUNDING) >> PRECISION)
+                        .clamp(0, max_value);
+                    let b = ((y_value + cb_coef * cb_value + ROUNDING) >> PRECISION)
+                        .clamp(0, max_value);
+                    let g = ((y_value - g_coef_1 * cr_value - g_coef_2 * cb_value + ROUNDING)
+                        >> PRECISION)
+                        .clamp(0, max_value);
+
+                    if CHANNELS == 4 {
+                        rgb_dst[0] = r.as_();
+                        rgb_dst[1] = g.as_();
+                        rgb_dst[2] = b.as_();
+                        rgb_dst[3] = max_value.as_();
+                    } else if CHANNELS == 3 {
+                        rgb_dst[0] = r.as_();
+                        rgb_dst[1] = g.as_();
+                        rgb_dst[2] = b.as_();
+                    } else {
+                        unreachable!();
+                    }
+                }
+            }
+        }
+    }
+
+    // Process remainder if height is odd
+
+    let y_iter = y_plane
+        .chunks_exact(y_stride * 2)
+        .remainder()
+        .chunks_exact(y_stride);
+    let rgb_iter = rgb.chunks_exact_mut(rgb_stride).rev();
+    let u_iter = u_plane.chunks_exact(u_stride).rev();
+    let v_iter = v_plane.chunks_exact(v_stride).rev();
+
+    for (((y_src, u_src), v_src), rgb) in y_iter.zip(u_iter).zip(v_iter).zip(rgb_iter) {
+        let y_iter = y_src.chunks_exact(2);
+        let rgb_chunks = rgb.chunks_exact_mut(CHANNELS * 2);
+        for (((y_src, u_src), v_src), rgb_dst) in y_iter.zip(u_src).zip(v_src).zip(rgb_chunks) {
+            let y_value = (y_src[0].as_() - bias_y) * y_coef;
+            let cb_value = u_src.as_() - bias_uv;
+            let cr_value = v_src.as_() - bias_uv;
+
+            let r = ((y_value + cr_coef * cr_value + ROUNDING) >> PRECISION).clamp(0, max_value);
+            let b = ((y_value + cb_coef * cb_value + ROUNDING) >> PRECISION).clamp(0, max_value);
+            let g = ((y_value - g_coef_1 * cr_value - g_coef_2 * cb_value + ROUNDING) >> PRECISION)
```

Comment:
The re-occurence of this (across all formats) makes me consider if there should be a helper:

```suggestion
#[inline(always)]
fn round<const PRECISION: i32, const MAX: i32>(val: i32) -> i32 {
    let ROUNDING: i32 = 1 << (PRECISION - 1);
    ((val + ROUNDING) >> PRECISION).clamp(0, MAX)
}

            let r = round::<PRECISION, MAX>(y_value + cr_coef * cr_value);
            let b = round::<PRECISION, MAX>(y_value + cb_coef * cb_value);
            let g = round::<PRECISION, MAX>(y_value - g_coef_1 * cr_value - g_coef_2 * cb_value);
```

The control flow / arithmetic order will be more apparent and it should not influence codegen.

Reply:
@awxkee Sorry if this was unclear, this was not meant as a literal suggestion but a rough draft of a refactor of these rounding conversions. Would you move that function out of the body and apply it to all the other occurences in the various yuv functions where it can be used? Maybe you're already doing so in the background.

Reply:
Yup, sure, no worries. This was clear enough.